### PR TITLE
ADNS-5050 / Ploopy Nano / Ploopy Mini Trackballs

### PR DESCRIPTION
**Pull request by ploopyco**
## Description

Added a driver for the ADNS-5050 optical mouse sensor.
Added implementations for the Ploopy Nano and Ploopy Mini trackball.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Review by drashna — APPROVED** [review]

**Comment by ploopyco** [comment]
Lint seems unhappy with the fact that keyboards/ploopyco/trackball_nano has no buttons (i.e. "layouts" is undefined in info.json and LAYOUT function is empty).

Code compiles and runs correctly, but I'm not sure how to fix the lint errors. Any ideas?

**Comment by ploopyco** [comment]
Testing identified a spurious scrolling issue. Will re-open PR later.

File: keyboards/ploopyco/trackball_nano/rev1_001/config.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * Keyboard Matrix Assignments
+ *
+ * Change this to how you wired your keyboard
+ * COLS: AVR pins used for columns, left to right
+ * ROWS: AVR pins used for rows, top to bottom
+ * DIODE_DIRECTION: COL2ROW = COL = Anode (+), ROW = Cathode (-, marked on diode)
+ *                  ROW2COL = ROW = Anode (+), COL = Cathode (-, marked on diode)
+ *
+ */
+#define DIRECT_PINS {}
```
Comment (drashna) [inline keyboards/ploopyco/trackball_nano/rev1_001/config.h:32]:
```suggestion
#define DIRECT_PINS { { NO_PIN } }
```

File: keyboards/ploopyco/trackball_nano/trackball_nano.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#define OPT_ENC2 F4
+#define OPT_ENC1_MUX 0
+#define OPT_ENC2_MUX 4
+
+void process_mouse(report_mouse_t* mouse_report);
+void process_mouse_user(report_mouse_t* mouse_report, int16_t x, int16_t y);
+void process_wheel(report_mouse_t* mouse_report);
+void process_wheel_user(report_mouse_t* mouse_report, int16_t h, int16_t v);
+
+#define LAYOUT() {}
```
Comment (drashna) [inline keyboards/ploopyco/trackball_nano/trackball_nano.h:39]:
```suggestion
#define LAYOUT() { { } }
```

**Review by drashna — COMMENTED** [review]
This may work better

File: keyboards/ploopyco/trackball_nano/rev1_001/config.h (hunk 1 above)
Comment (ploopyco) [reply keyboards/ploopyco/trackball_nano/rev1_001/config.h:32]: Tried both of these. Unfortunately, lint error persists, same as before.

Could commit these suggestions anyways if you think it'll help.

**Comment by drashna** [comment]
Should be fine.  Lint may not be happy, but it compiles, so that's what matters.

File: keyboards/ploopyco/trackball_mini/readme.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
+
+# Ploopy Trackball Mini
+
+![Ploopyco Trackball Mini](TODO stick a link in here)
```
Comment (drashna) [inline keyboards/ploopyco/trackball_mini/readme.md:4]: Needs an image.

File: keyboards/ploopyco/trackball_mini/readme.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+
+* Maintainer: [PloopyCo](https://github.com/ploopyco)
+* Key contributors: [Drashna Jael're](https://github.com/drashna/), [Germ](https://github.com/germ/)
+* Hardware Supported: ATMega32u4 16MHz(5v)  
+* Hardware Availability: [Store](https://ploopy.co), [GitHub](https://github.com/ploopyco)
+
+Make example for this trackball (after setting up your build environment):
+
+    make ploopyco/trackball/rev1_001:default:flash
+    make ploopyco/trackball/rev1_001:via:flash
```
Comment (drashna) [inline keyboards/ploopyco/trackball_mini/readme.md:16]:
```suggestion
    make ploopyco/trackball_mini:default:flash
    make ploopyco/trackball_mini:via:flash
```

File: keyboards/ploopyco/trackball_mini/trackball_mini.c [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+void pointing_device_task(void) {
+    report_mouse_t mouse_report = pointing_device_get_report();
+    process_wheel(&mouse_report);
+    process_mouse(&mouse_report);
+
+    pointing_device_set_report(mouse_report);
+    if (has_report_changed(mouse_report, pointing_device_get_report())) {
+        pointing_device_send();
+    }
+}
```
Comment (drashna) [inline keyboards/ploopyco/trackball_mini/trackball_mini.c:225]: This logic is built into `pointing_device_send()` by default now. Can be slimmed down to: 
```suggestion
void pointing_device_task(void) {
    report_mouse_t mouse_report = pointing_device_get_report();
    process_wheel(&mouse_report);
    process_mouse(&mouse_report);

    pointing_device_set_report(mouse_report);
    pointing_device_send();
}
```

File: keyboards/ploopyco/trackball_mini/trackball_mini.c [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+
+void eeconfig_init_kb(void) {
+    keyboard_config.dpi_config = PLOOPY_DPI_DEFAULT;
+    eeconfig_update_kb(keyboard_config.raw);
+}
```
Comment (drashna) [inline keyboards/ploopyco/trackball_mini/trackball_mini.c:230]:
```suggestion
	eeconfig_init_user();
}
```

File: keyboards/ploopyco/trackball_nano/trackball_nano.c [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+    pointing_device_set_report(mouse_report);
+    if (has_report_changed(mouse_report, pointing_device_get_report())) {
+        pointing_device_send();
+    }
+}
+
+void eeconfig_init_kb(void) {
+    keyboard_config.dpi_config = PLOOPY_DPI_DEFAULT;
+    eeconfig_update_kb(keyboard_config.raw);
+}
```
Comment (drashna) [inline keyboards/ploopyco/trackball_nano/trackball_nano.c:204]:
```suggestion

void pointing_device_task(void) {
    report_mouse_t mouse_report = pointing_device_get_report();
    process_wheel(&mouse_report);
    process_mouse(&mouse_report);

    pointing_device_set_report(mouse_report);
    pointing_device_send();
}

void eeconfig_init_kb(void) {
    keyboard_config.dpi_config = PLOOPY_DPI_DEFAULT;
    eeconfig_update_kb(keyboard_config.raw);
	eeconfig_init_user();
}
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/ploopyco/trackball_nano/info.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
+    "keyboard_name": "Ploopy Trackball Nano",
+    "url": "www.ploopy.co",
+    "maintainer": "ploopyco",
+    "manufacturer": "Ploopy Corporation",
+}
```
Comment (drashna) [inline keyboards/ploopyco/trackball_nano/info.json:6]: Horrible hack, but it gets past lint. 
```suggestion
     "layouts": {
        "LAYOUT": {
            "layout": [
                {"x":0, "y":0}
            ]
        }
    }
}
```

**Review by drashna — APPROVED** [review]

**Review by tzarc — APPROVED** [review]

**Pull request merged**